components/style/traversal.rs

```diff
@@ -113,24 +132,113 @@ pub trait DomTraversalContext<N: TNode> {
     /// performed.
     ///
     /// If it's false, then process_postorder has no effect at all.
-    fn needs_postorder_traversal(&self) -> bool { true }
+    fn needs_postorder_traversal() -> bool { true }
 
-    /// Returns true if traversal should visit the given child.
-    fn should_traverse_child(child: N) -> bool;
+    /// Must be invoked before traversing the root element to determine whether
+    /// a traversal is needed. Returns a token that allows the caller to prove
+    /// that the call happened.
+    ///
+    /// The skip_root parameter is used in Gecko to style newly-appended children
+    /// without restyling the parent.
+    fn pre_traverse(root: N::ConcreteElement, stylist: &Stylist, skip_root: bool)
+                    -> PreTraverseToken
+    {
+        // If we should skip the root, traverse unconditionally.
+        if skip_root {
+            return PreTraverseToken {
+                traverse: true,
+                skip_root: true,
+            };
+        }
+
+        // Expand the snapshot, if any. This is normally handled by the parent, so
+        // we need a special case for the root.
+        //
+        // Expanding snapshots here may create a LATER_SIBLINGS restyle hint, which
+        // we will drop on the floor. This is fine, because we don't traverse roots
+        // with siblings.
+        debug_assert!(root.next_sibling_element().is_none());
+        if let Some(mut data) = root.mutate_data() {
+            if let Some(r) = data.as_restyle_mut() {
+                let _later_siblings = r.expand_snapshot(root, stylist);
+            }
+        }
+
+        PreTraverseToken {
+            traverse: Self::node_needs_traversal(root.as_node()),
+            skip_root: false,
+        }
+    }
+
+    /// Returns true if traversal should visit a text node. The style system never
+    /// processes text nodes, but Servo overrides this to visit them for flow
+    /// construction when necessary.
+    fn text_node_needs_traversal(_: N) -> bool { false }
```

Comment:
Please assert that the node is a text node here. That way if we change the filter in `children`, we can catch it easily (though there may be other assertions that could catch that?).